### PR TITLE
HBHE = Extra SOI fix

### PR DESCRIPTION
**Pull request by mariadalfonso**
removed more hard coded SOI in view of the TS10->TS8
discussion triggered by #21651

commit 57507f2 take the threshold from the GT instead of of 2.7 from the python config.
https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HcalRecProducers/python/HBHEMethod3Parameters_cfi.py#L6

i.e. 2.7 was left over from 2015 (3*HPD noise)

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@kpedro88 
please check if the valgrind error goes away here.

The PR is much less trivial than a simple fix to the bad memory reads.

**Comment by kpedro88** [comment]
@slava77 I'll check - this PR addresses all the places we noticed that hadn't been modified to allow for different # TS, not just the bad memory read noticed before

**Comment by kpedro88** [comment]
@slava77 I confirm the valgrind error goes away.

**Comment by perrotta** [comment]
@mariadalfonso : jenkins tests show changes in HBHERecHitsSorted for most of the workflows (including run1). I was thinking at 57507f2 , but if I understand correctly what you wrote it should be uneffective on run1 workflows.

For run1 it could be that https://github.com/cms-sw/cmssw/pull/21730/files#diff-2c377061097919e5196e1f3b0bae144bR33 makes the difference.

Instead of guessing: do you have an explanation for all those changes?

File: RecoLocalCalo/HcalRecAlgos/src/PedestalSub.cc [hunk 1]
```diff
@@ -17,28 +17,28 @@ void PedestalSub::init(int runCond=0, float threshold=0.0, float quantile=0.0) {
...
-    for (auto i=0; i<10; i++) {
-      if (i==4||i==5) continue;
-      if ( (inputCharge[i]-inputPedestal[i])<fThreshold) {
+    for (auto i=0; i<nTS; i++) {
+      if (i==soi||i==soi+1) continue;
+      if ( (inputCharge[i]-inputPedestal[i])<3*inputNoise[i]) {
 	baseline+=(inputCharge[i]-inputPedestal[i]);
       }
       else {
 	baseline+=fThreshold;
```
Comment (perrotta) [inline RecoLocalCalo/HcalRecAlgos/src/PedestalSub.cc]: is ```fThreshold``` still ok here?

**Comment by mariadalfonso** [comment]
@perrotta 
The M3 is only correct for the 25ns, and will not give sensible results for the  50ns.

But with this 10_X release, we run the 25ns reco for all periods (see slava's comment
https://github.com/cms-sw/cmssw/pull/20986#issuecomment-338926012) so we should see some difference also there. Especially because there was the QPLL problem

File: RecoLocalCalo/HcalRecAlgos/src/PedestalSub.cc (hunk 1 above)
Comment (mariadalfonso) [reply RecoLocalCalo/HcalRecAlgos/src/PedestalSub.cc]: @perrotta 
indeed fThreshold should have been fixed also here

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
At this point, the PedestalSub class could be simplified, as none of the class members is needed:
```
  float fThreshold;
  float fQuantile;
  float fCondition;
```
as well as the method
```
  void init(int runCond, float threshold, float quantile);
```
(also in a later cleaning PR, if we want to merge this quickly for pre3)

**Comment by mariadalfonso** [comment]
@perrotta 
I will take note of this extra code cleaning, will make a separate PR together with other code cleanup in other places.

Indeed it's important to fix completely the TS10->TS8 for the pre3

**Comment by perrotta** [comment]
Jenkins tests show many more differences now, i.e. they propagate to jets: is this due to chenge of the pedestal threshold?
@mariadalfonso could you please quickly investigate and report?
(Perhaps. given the urgency of applying the other corrections that remove hardcoded values it could have been wiser to make two separate pull requests... but ok, just check the origin of the new differences and let us know)

**Comment by mariadalfonso** [comment]
@perrotta 

You should see minor effects on the HLT jets and mainly on the siPM, but I do not expect changes on the offline jets.

From various printout the typical noise 3*inputNoise[i] from the GT is
2.7-3 fC for the HPD and is ~ 65 fC for the siPM.

This is roughly 0.5 GeV for the HPD 0.03 GeV for the siPM

The gain (fC->GeV) is ~ 5 for HPD and ~ 2000 for siPM
https://github.com/cms-sw/cmssw/blob/CMSSW_10_0_0_pre2/CalibCalorimetry/HcalPlugins/python/Hcal_Conditions_forGlobalTag_cff.py#L114
https://github.com/cms-sw/cmssw/blob/CMSSW_10_0_0_pre2/CalibCalorimetry/HcalPlugins/python/Hcal_Conditions_forGlobalTag_cff.py#L36

**Comment by perrotta** [comment]
@mariadalfonso 

Differences observed in the jekins outputs are small, but nonetheless present, see for example 2018 workflow TTbar13TeV2018 wf10824
image

If you can't explain it, this certainly deserves some deeper investigation. In that case, I would suggest you to remove the threshold fixes from this PR and move to a separate one that can be investigated with calm. If the current one only contains the removal of hardcoded parametet, which is needed now, it can be approved at once.

**Comment by mariadalfonso** [comment]
@perrotta 

Changes in the offline "M2" recHit (only for energy ~< 20GeV) comes from the fix from the first commit in this PR
https://github.com/cms-sw/cmssw/pull/21730/commits/d13fb7cf9e42d2892170a271296985dccae06484

**Comment by abdoulline** [comment]
@mariadalfonso  Maria, may be I'm missing something, but this fix

https://github.com/cms-sw/cmssw/pull/21730/files#diff-32305d93865b95ded358d577d024722bR107

slightly changes M2 (now that we have 8TS default), right ?

**Comment by abdoulline** [comment]
@mariadalfonso    Ah, I was too slow, you've already answered before I've asked.

**Comment by perrotta** [comment]
@mariadalfonso  @abdoulline 

It is expected that there are changes in recHits due to d13fb7c

What I noticed is that before the last commit we had:
- Reco comparison results: 291 differences found in the comparisons
while after that last commit this reads now
- Reco comparison results: 439 differences found in the comparisons
and the extra differences are from jets, pf candidates etc.

My point is: is it understood the origin of this difference before and after the last commit?

If thresholds for pedestald are touched, I wouldn't be surprised to see effects downhill. But if you say that this is not expected, I think some deeper check is needed to understand why.

**Comment by abdoulline** [comment]
@perrotta  sorry for my misunderstanding. 
Do I understand it right (now...) that after the last commit  https://github.com/cms-sw/cmssw/commit/1aaf14a10c6f3391d48cc52f7e40a122b53411cb
some offline quantities got affected and the plot you refer to this morning illustrated the difference in offline (PF jets) after this commit?

**Comment by perrotta** [comment]
Yes Salavat, that is what puzzles me, given that Maria claims that no  
differences should be expected.

Salavat Abdullin <notifications@github.com> ha scritto:

**Comment by abdoulline** [comment]
OK, I see..  Actually I didn't find any usage of the updated file (PedestalSub::calculate) other 
than in HcalDeterministicFit.cc (=M3) several days ago (in the context of 
this PR) and just have been re-checked again with the same result...

**Comment by mariadalfonso** [comment]
@perrotta 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21730/25128/summary.html
Reco comparison results: 291 differences found in the comparisons
IB is CMSSW_10_0_X_2017-12-14-2300

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21730/25162/summary.html
Reco comparison results: 439 differences found in the comparisons
IB is CMSSW_10_0_X_2017-12-18-1500

there are differences in the IB in the HCAL side due to the changes in the pulse shape for siPM that became active with this PR
https://github.com/cms-sw/cmssw/pull/21689

Can the difference between the number 291 and  439 due to other M2 changes in other PR?

**Comment by deguio** [comment]
Hi @perrotta 
I've run the following test:
- used CMSSW_10_0_X_2017-12-19-1100 as starting release
- plug on top the first three commits of this PR --> my reference
- plug on top the whole PR (4 commits)
- run WF 10824.0 from runTheMatrix in the two cases
- compare the jet distribution from step3.root

and I didn't see any difference. I'm attaching the comparison.
screen shot 2017-12-19 at 14 04 28

**Comment by perrotta** [comment]
@deguio : thank you for checking. It seems to me that yuor plots, in spite of having the same 114 Jets from the first 10 events processed with 10824.0, has a different binning, and therefore the different event observed by jenkins could even have remained inside the same bin in your plot.

I don't think that comparing with different IBs can be the source of the difference by itself: if you compare wrt baseline, then the difference can only be due to what you add on top of it.

It is clear that there is some non reproducibility beyond: such a non reproducibility can be in the baseline itself, and in fact this is what this PR wants to fix. 

My proposal is to merge this PR and see if the non reproducibility issues disappear with it. We have the check of Kevin,  https://github.com/cms-sw/cmssw/pull/21730#issuecomment-352460863, that shows how the valgrind errors go away with it.

With this caveat in mind, I'm going to sign this PR. I think this will be discussed today at the ORP (@davidlange6 @slava77) at which I will unfortunately be impossibilitate to connect.

**Comment by perrotta** [comment]
+1
- SOI and number of samples in HCal local reco not hardcoded any more
- Fix in pedestals threshold determination
- Valgrind erros go away (thank you Kevin for checking)
- Small regression in HBHE recHits, as expected
- Some other regressions visible in jets, pfcandidates, etc. could be due to non reproducibility issues already present in the baseline IB: it is worth checking if they disappear once this PR is merged

**Comment by deguio** [comment]
thanks @perrotta 
sounds good

**Comment by davidlange6** [comment]
+1

**Pull request merged**